core/src/apps/cardano/sign_tx.py

```diff
@@ -4,238 +4,181 @@
-                ]
-            )
+def _serialize_tx(keychain: seed.Keychain, msg: CardanoSignTx) -> Tuple[bytes, bytes]:
+    tx_body = _build_tx_body(keychain, msg)
+    tx_hash = _hash_tx_body(tx_body)
 
-        return witnesses
+    witnesses_for_cbor = _build_witnesses(keychain, msg.inputs, tx_hash)
+    witnesses = {0: _detupleize(witnesses_for_cbor)}
 
-    @staticmethod
-    def compute_fee(input_coins_sum: int, outgoing_coins: list, change_coins: list):
-        outgoing_coins_sum = sum(outgoing_coins)
-        change_coins_sum = sum(change_coins)
+    serialized_tx = cbor.encode([tx_body, witnesses, None])
 
-        return input_coins_sum - outgoing_coins_sum - change_coins_sum
+    return serialized_tx, tx_hash
 
-    def serialise_tx(self):
 
-        self._process_outputs()
+def _build_tx_body(keychain: seed.Keychain, msg: CardanoSignTx) -> Dict:
+    inputs_for_cbor = _build_inputs(msg.inputs)
+    outputs_for_cbor = _build_outputs(keychain, msg.outputs, msg.protocol_magic)
 
-        inputs_cbor = []
-        for input in self.inputs:
-            inputs_cbor.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([input.prev_hash, input.prev_index])),
-                ]
-            )
+    tx_body = {
+        0: _detupleize(inputs_for_cbor),
+        1: _detupleize(outputs_for_cbor),
+        2: msg.fee,
+        3: msg.ttl,
+    }
 
-        inputs_cbor = cbor.IndefiniteLengthArray(inputs_cbor)
+    return tx_body
 
-        outputs_cbor = []
-        for index, address in enumerate(self.output_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.outgoing_coins[index]]
-            )
 
-        for index, address in enumerate(self.change_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.change_coins[index]]
+def _build_inputs(inputs: List[CardanoTxInputType]) -> List[Tuple[bytes, int]]:
+    result = []
+    for input in inputs:
+        result.append((input.prev_hash, input.prev_index))
+
+    return result
+
+
+def _build_outputs(
+    keychain: seed.Keychain, outputs: List[CardanoTxOutputType], protocol_magic: int
+) -> List[Tuple[bytes, int]]:
+    result = []
+    total_amount = 0
+    for output in outputs:
+        amount = output.amount
+        if output.address_n:
+            address, _ = derive_address_and_node(
+                keychain, output.address_n, protocol_magic
             )
+        else:
+            address = output.address
+            if address is None:
+                raise wire.ProcessError(
+                    "Each output must have address or address_n field!"
+                )
+            if not is_safe_output_address(address):
+                raise wire.ProcessError("Invalid output address!")
+
+        total_amount += amount
+        result.append((base58.decode(address), amount))
```

Comment:
We should make sure using base58 is compatible with trezor connect --- I have no idea. (I would prefer to receive bytes, but we can change it later in another shelley PR).

Reply:
It is compatible. It has worked the same way in Byron. Yes, we will be receiving bytes and it will be done in the second PR since it doesn't make sense here yet.